src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs

```diff
@@ -606,5 +615,38 @@ private ProcessWaitState GetWaitState()
         public bool Responding => true;
 
         private bool WaitForInputIdleCore(int milliseconds) => throw new InvalidOperationException(SR.InputIdleUnkownError);
+
+        private static bool s_sigchildHandlerRegistered = false;
+        private static readonly object s_sigchildGate = new object();
+        private static readonly Interop.Sys.SigChldCallback s_sigChildHandler = OnSigChild;
+
+        private static bool EnsureSigChildHandler()
+        {
+            lock (s_sigchildGate)
+            {
+                if (!s_sigchildHandlerRegistered)
+                {
+                    if (!Interop.Sys.InitializeSignalHandling())
+                    {
+                        throw new Win32Exception();
+                    }
+
+                    Interop.Sys.RegisterForSigChld(s_sigChildHandler);
+
+                    s_sigchildHandlerRegistered = true;
+                }
+            }
+            return s_sigchildHandlerRegistered;
+        }
+
+        private static void OnSigChild()
+        {
+            // Lock to avoid races with Process.Start
+            lock (s_sigchildGate)
+            {
+                ProcessWaitState.CheckChildren();
+                Interop.Sys.ResumeSigChld();
```

Comment:
Why does managed code need to call this?  Why isn't it just called from the native code after it invokes the callback?  Then this wouldn't need to be exposed as a P/Invoke at all right?

Reply:
When disposition was SIG_IGN, ResumeSigChld is reaping all remaining children, so it is here under s_sigchildGate.
If the original handler gets called on the signal thread, we can implement this in c#.